crates/walrus-sui/src/client/retry_client/download_handler.rs

```diff
@@ -0,0 +1,606 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements a checkpoint download handler for the RPC client.
+use std::{
+    sync::{
+        Arc,
+        atomic::{AtomicU64, AtomicUsize, Ordering},
+    },
+    time::Instant,
+};
+
+use atomic_time::AtomicInstant;
+
+use crate::client::{retry_client::RetriableClientError, rpc_config::RpcFallbackConfig};
+
+/// The decision to make where to download a checkpoint.
+#[derive(Debug)]
+pub enum CheckpointDownloadDecision {
+    /// Attempt primary RPC first
+    AttemptPrimary,
+    /// Try fallback if primary fails
+    TryFallback(Option<RetriableClientError>),
+    /// Use primary error if fallback fails
+    UsePrimaryError(RetriableClientError),
+}
+
+/// Updated failure stats for the current sequence number
+#[derive(Debug)]
+pub struct CheckpointFailureStats {
+    /// Overall number of failures since last_success
+    /// Note: This is not the number of failures for the current sequence number
+    /// This is the number of failures since the last successful RPC call
+    overall_failures_count: usize,
+    /// Number of consecutive failures for the current sequence number
+    consecutive_failures_for_this_seq: usize,
```

Comment:
Suggestion: IIUC, say upon 3 failures, a single checkpoint will fallback to archival, if there are consecutive failures, second may need to fail 6 times, and 9 may need to failre 9 times to fallback. This is a bit unintuitive.

What do you think of tracking BTreeMap<SequenceNumber, NumberOfFailure>? In this way, this also gives a more clear picture what's the lowest failure seq, and what's the failure count for each seq. Given that the number of checkpoint downloader is limited, the size of thie BTreeMap is also limited.

Reply:
I'm fine with storing state per failing checkpoint sequence number since the total number of inflight checkpoints is never going to be too much. We definitely need to wrap the `BTreeMap` with a mutex since we will be concurrently reading/writing from it. I think that removes the need to do any atomics as well which simplifies things a bit as well.